protocols/kad/src/bootstrap.rs

```diff
@@ -0,0 +1,169 @@
+use futures::FutureExt;
+use std::task::{Context, Poll, Waker};
+use std::time::Duration;
+
+use futures_timer::Delay;
+
+use crate::QueryStats;
+
+#[derive(Debug)]
+pub(crate) struct Status {
+    interval: Option<Duration>,
+    next_periodic_bootstrap: Option<Delay>,
+    waker: Option<Waker>,
+
+    bootstrap_asap: bool,
+
+    is_bootstrapping: bool,
+    any_bootstrap_succeeded: bool,
+}
+
+impl Status {
+    pub(crate) fn new(interval: Option<Duration>) -> Self {
+        Self {
+            interval,
+            next_periodic_bootstrap: interval.map(Delay::new),
+            waker: None,
+            bootstrap_asap: false,
+            is_bootstrapping: false,
+            any_bootstrap_succeeded: false,
+        }
+    }
+
+    pub(crate) fn on_new_peer_in_routing_table(&mut self) {
+        self.bootstrap_asap = true;
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake()
+        }
+    }
+
+    pub(crate) fn on_started(&mut self) {
+        self.is_bootstrapping = true;
+    }
+
+    pub(crate) fn on_result(&mut self, stats: &QueryStats) {
+        if stats.num_successes() > 0 {
+            self.any_bootstrap_succeeded = true;
+        }
+
+        self.is_bootstrapping = false;
+    }
+
+    pub(crate) fn poll_next_bootstrap(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        if self.bootstrap_asap {
+            self.bootstrap_asap = false;
+            if let (Some(interval), Some(delay)) =
+                (self.interval, self.next_periodic_bootstrap.as_mut())
+            {
+                delay.reset(interval);
+            }
+
+            return Poll::Ready(());
+        }
+
+        if let (Some(interval), Some(delay)) =
```

Comment:
Since `interval` and `next_periodic_bootstrap` are tightly coupled together, would it be better to have them in a single field `Option<(Duration, Delay)>` ?

Don't really have an opinion on this.

Reply:
Good idea!